src/Database/LSMTree/Internal/Run/Index/Compact.hs

```diff
@@ -1,21 +1,506 @@
+  , ciPrimary              :: !(VU.Vector Word32)
+
+    -- | A clash on page @i@ means that the minimum key of page @i@ and the
+    -- maximum key of the predecessor page @i-1@ have the exact same @32@
+    -- primary bits. Use the tie-breaker index to break ties.
+    --
+    -- TODO: we could use a ByteArray here instead
+  , ciClashes              :: !(VU.Vector Bit)
+    -- | Maps a page @i@ to its full minimum key, but only if there is a clash
+    -- on page @i@.
+    --
+    -- INVARIANT: a key is included in this 'IntMap' if and only if 'ciClashes'
+    -- recorded that there is a class. In code: @isJust (IntMap.lookup ix
+    -- ciTiebreaker) == ciClashes VU.! ix@.
+  , ciTieBreaker           :: !(IntMap k)
+  }
+
+{-------------------------------------------------------------------------------
+  Invariants
+-------------------------------------------------------------------------------}
+
+-- | Inclusive bounds for range-finder bit-precision.
+rangeFinderPrecisionBounds :: (Int, Int)
+rangeFinderPrecisionBounds = (0, 16)
+
+-- | https://en.wikipedia.org/wiki/Birthday_problem#Probability_of_a_shared_birthday_(collision)
+suggestRangeFinderPrecision :: Int -> Int
+suggestRangeFinderPrecision maxPages =
+    clamp $ ceiling @Double $ 2 * logBase 2 (fromIntegral maxPages)
+  where
+    (lb, ub)            = rangeFinderPrecisionBounds
+    clamp x | x < lb    = lb
+            | x > ub    = ub
+            | otherwise = x
+
+{-------------------------------------------------------------------------------
+  Queries
+-------------------------------------------------------------------------------}
+
+-- | Given a search key, find the number of the disk page that /might/ contain
+-- this key.
+--
+-- Note: there is no guarantee that the disk page contains the key. However, it
+-- is guaranteed that /no other/ disk page contain the key.
+--
+search ::
+     forall k. (SliceBits k, Integral k)
+  => k
+  -> CompactIndex k
+  -> Maybe Int
+search k CompactIndex{..} = runST $ do
+    let lb = ciRangeFinder VU.! fromIntegral (topBits ciRangeFinderPrecision k)
+        ub = ciRangeFinder VU.! fromIntegral (topBits ciRangeFinderPrecision k + 1)
+    -- We have to thaw the primary index to use the vector search algorithm. It is
+    -- safe to use 'unsafeThaw' here because, once frozen, a compact index can not
+    -- be manipulated anymore.
+    !ciPrimary' <- VU.unsafeThaw ciPrimary
+    -- Compute @pageNr@ similar to 'IntMap.lookupLE': find the page number
+    -- @pageNr@ of the the last entry in the primary array (within bounds @[lb,
+    -- ub]@) that is smaller or equal to our search key.
+    --
+    -- Regardless of whether the search key is actually present in a page, it
+    -- /can not/ exist in a page @>pageNr@.
+    pageNr' <- VA.gallopingSearchLeftPBounds
+                (> sliceBits32 ciRangeFinderPrecision 32 k)
+                ciPrimary'
+                (fromIntegral lb)
+                (fromIntegral ub)
```

Comment:
The only bit that really needs to be in ST is this bit. It would be slightly clearer to use runST on just this bit.